contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,433 @@
+to the Objects that represent the target Object state may cause the serialized 
+representation of those Objects to be unequal even when they are semantically 
+equivalent.
+
+The controller should deserialize the values of the AnyStates representing their 
+target Object state and perform a deep, semantic equality test. Here all 
+differences that do not constitute a mutation to the target Object state are 
+disregarded during the equivalence test.
+    
+### Target Object State Reconciliation
+There are three ways for a controller to reconcile a generated Object with the 
+declared target Object state.
+
+1. If the target Object state is [equivalent](#version-equivalence) to the 
+target Object state associated with the generated Object, the controller will 
+update the associated [version tracking information](#version-tracking).
+1. If the Object can be updated in place to reconcile its state with the 
+current target Object state, a controller may update the Object in place 
+provided that the associated version tracking information is updated as well.
+1. Otherwise, the controller must destroy the Object and recreate it from the 
+current target Object state.
+
+## Kubectl
+
+Modifications to kubectl to leverage controller history are an optional 
+extension. Users can trigger rolling updates and rollbacks by modifying their 
+manifests and using `kubectl apply`. Controllers will be able to detect 
+revisions to their target Object state and perform 
+[reconciliation](#target-object-state-reconciliation) as necessary.
+
+### Viewing History 
+
+Users can view a controller's revision history with the following command.
+
+```bash
+> kubectl rollout history
+```
+
+To view the details of the revision indicated by `<revision>`. Users can use 
+the following command.
+
+```bash
+> kubectl rollout history --revision <revision>
+```
+
+### Rollback
+
+For future work, `kubeclt rollout undo` can be implemented in the general case 
+as an extension of the [above](#viewing-history ).
+
+```bash
+> kubectl rollout undo
+```
+
+Here `kubectl undo` simply uses strategic merge patch to apply the state 
+contained at a particular revision. 
+
+## Tests
+
+1. Controllers can create an AnyState containing a revision of their target 
+Object state.
+1. Controllers can reconstruct their revision history.
+1. Controllers can't update an AnyState's `.Data`.
+1. Controllers can delete an AnyState to maintain their history with respect to 
+the configured `RevisionHistoryLimit`.
+
+## Appendix
+ 
+### Hashing
+We will require a CRHF (collision resistant hash function), but, as we expect 
+no adversaries, such a function need not be resistant to pre-image and 
+secondary pre-image attacks.
+As the property of interest is primarily collision resistance, and as we 
+provide a method of [collision resolution](#collision-resolution), both 
+cryptographically strong functions, such as Secure Hash Algorithm 2 (SHA-2), 
```

Comment:
Based on prior experimentation, I think the gain we get from the difference in stability between sha-2 and fnv is negligible because fnv is already stable enough. Coupled with the uniquifier, we should be foolproof.